### PR TITLE
Added TransD

### PR DESCRIPTION
**Pull request by samboy83**
Implemented TransD model
Added TransD to unittest

### CODE REVIEW

**Comment by tillgeissler** [comment]
I don't know if it is necessary, but for TransH, they implemented the penalty corresponding to the constraint proposes in the paper. Maybe we should also include the constraint on the projection vectors as penalty

**Comment by samboy83** [comment]
I'm not sure neither, but I think we can control that constraints by applying lookup_embedder.normalize.p: 1. or 2. like TransE.
This is my guess, the additional penalty part of TransH seems to have been implemented because its author used different cost function with the additional penalty term in their paper.

File: kge/model/transd.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    @staticmethod
+    def _transfer(ent_emb, ent_norm_vec_emb, rel_norm_vec_emb):
+        d = ent_norm_vec_emb.size(1)
+        k = rel_norm_vec_emb.size(1)
+
+        min_dim = min(d, k)
+
+        out = rel_norm_vec_emb * torch.sum(ent_norm_vec_emb * ent_emb, dim=-1, keepdim=True)
+        out[:, :min_dim] += ent_emb[:, :min_dim]
```
Comment (LJ171) [inline kge/model/transd.py]: Your previous implementation was closer to the original formula from the paper but I guess this change produces the same results right?

File: kge/model/transd.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+transd:
+  class_name: TransD
+  entity_embedder:
+    type: lookup_embedder
+    +++: +++
+  relation_embedder:
+    type: lookup_embedder
+    +++: +++
+
+  l_norm: 1.
```
Comment (LJ171) [inline kge/model/transd.yaml:12]: In the paper they always use l_norm = 2, would it make sense to set that as the default value?

**Review by LJ171 — COMMENTED** [review]
Sorry for the late review, taught I already have done that but must have forgotten to click 'Finish your review'. LGTM only got two minor questions.

**Comment by LJ171** [comment]
> I'm not sure neither, but I think we can control that constraints by applying lookup_embedder.normalize.p: 1. or 2. like TransE.
> This is my guess, the additional penalty part of TransH seems to have been implemented because its author used different cost function with the additional penalty term in their paper.

I don't think that it is necessary to implement the penalty corresponding to the constraint but I'm not sure either.

File: kge/model/transd.py (hunk 1 above)
Comment (samboy83) [reply kge/model/transd.py]: Yes, I checked whether these two formula produce the same results. and honestly I referred to pykeen and other libraries. My original method was too slow because it used matrix calculations. So I changed this efficiently not to use matrix calculation.

File: kge/model/transd.yaml (hunk 2 above)
Comment (samboy83) [reply kge/model/transd.yaml:12]: I think it makes more sense to use l_norm = 2. But the default value of the other models such as TransE, TransH and RotatE is l_norm = 1. So I just didn't touch it.

**Review by LJ171 — APPROVED** [review]
Since no one else seems to have comments about this PR. I would say it's done and you can open a second PR to the uma repo 👍
